src/GraphQL/Validation/Rules/NoUnusedFragments.cs

```diff
@@ -21,33 +21,41 @@ public class NoUnusedFragments : IValidationRule
 
         /// <inheritdoc/>
         /// <exception cref="NoUnusedFragmentsError"/>
-        public Task<INodeVisitor> ValidateAsync(ValidationContext context)
-        {
-            var operationDefs = new List<Operation>();
-            var fragmentDefs = new List<FragmentDefinition>();
+        public Task<INodeVisitor> ValidateAsync(ValidationContext context) => _nodeVisitor;
 
-            return new EnterLeaveListener(_ =>
+        private static readonly Task<INodeVisitor> _nodeVisitor = new NodeVisitors(
+            new MatchingNodeVisitor<Operation>((node, context) =>
             {
-                _.Match<Operation>(node => operationDefs.Add(node));
-                _.Match<FragmentDefinition>(node => fragmentDefs.Add(node));
-                _.Match<Document>(leave: document =>
+                (context.TypeInfo.NoUnusedFragments_OperationDefs ??= new List<Operation>(1)).Add(node);
+                context.TypeInfo.NoUnusedFragments_FragmentDefs?.Clear();
+            }),
+            new MatchingNodeVisitor<FragmentDefinition>((node, context) =>
+            {
+                var fragmentDefs = context.TypeInfo.NoUnusedFragments_FragmentDefs ??= new List<FragmentDefinition>();
+                fragmentDefs.Add(node);
+            }),
+            new MatchingNodeVisitor<Document>(leave: (document, context) =>
+            {
+                var fragmentDefs = context.TypeInfo.NoUnusedFragments_FragmentDefs;
+                if (fragmentDefs == null || fragmentDefs.Count == 0)
```

Comment:
```suggestion
                if (fragmentDefs == null)
```
Removed condition is impossible